pkg/tasks/local/event_feed.go

```diff
@@ -0,0 +1,62 @@
+package local
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/encoding/protojson"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	v1 "github.com/conductorone/baton-sdk/pb/c1/connectorapi/baton/v1"
+	"github.com/conductorone/baton-sdk/pkg/tasks"
+	"github.com/conductorone/baton-sdk/pkg/types"
+)
+
+type localEventFeed struct {
+	o sync.Once
+}
+
+func (m *localEventFeed) Next(ctx context.Context) (*v1.Task, time.Duration, error) {
+	var task *v1.Task
+	m.o.Do(func() {
+		task = &v1.Task{
+			TaskType: &v1.Task_EventFeed{},
+		}
+	})
+	return task, 0, nil
+}
+
+func (m *localEventFeed) Process(ctx context.Context, task *v1.Task, cc types.ConnectorClient) error {
+	var pageToken string
+	for {
+		resp, err := cc.ListEvents(ctx, &v2.ListEventsRequest{
+			PageSize:         100,
+			PageToken:        pageToken,
+			StartingPosition: task.GetEventFeed().GetStartingPosition(),
+		})
+		if err != nil {
+			return err
+		}
+		for _, event := range resp.GetEvents() {
+			bytes, err := protojson.Marshal(event)
+			if err != nil {
+				return err
+			}
+			//nolint:forbidigo
+			fmt.Println(string(bytes))
+		}
+		pageToken = resp.GetNextPageToken()
+		if pageToken == "" {
```

Comment:
I wonder if the semantics of a page token are exactly what we want. Normally we loop until the page token is empty, and then we know there isn't any data left. We may want to consider another field on the response to track the 'page token' to save for the next request. 

Imagine we return 10 results per page. On the 3rd page, we only have 5 results, so we'd return an empty page token. But the connector also needs to provide a token to use for the next call to resume from after the 5th result that it returned. That way when the connector is woken up in an hour, it can start off where it was at last more easily.

Reply:
To sum that up: since it is a feed, there will always be a next page token, so we can't use the same value to know that we need to continue making more requests. Maybe it is a bool on the response message?

Reply:
Well, it gets worse. Feeds are assumed to log data from oldest to newest, but in the two connectors that we are planning on implementing this for first, they return data from newest to oldest with no way of inverting it. We got around this in another connector by looping over all events it was able to find. If we hit a pageSize cutoff, it would return the next page token that the service needs and use that as the page token. If not, we returned the most recent timestamp and put a filter to only get objects newer than that timestamp. But this doesn't really work either because events in a service may come in out of order. We either have to live with that to a certain extend while minimizing our risk for it, or sync everything, or provide some ID tracking mechanism that says "well get everything after this day - a few and only return net new IDs." Lots of ideas on how to tackle this one. This should definitely be solved by the SDK to make connector implementation easier.